### PR TITLE
[NPU][Diffusion] Add SenseNova-U1 request batching and fused denoise optimizations

### PR DESCRIPTION
**Pull request by syd520zy**
## Motivation

This PR follows the SenseNova-U1 support and performance roadmap tracked in [#37742](https://github.com/sgl-project/sglang/issues/37742).

SenseNova-U1 previously processed serving requests one at a time on Ascend NPU. Simply enabling request batching provided almost no throughput improvement because padded SDPA attention, repeated normalization and MLP launches, KV layout conversions, and repeated timestep embedding work offset the batching benefit.

This change adds homogeneous request-level batching for SenseNova-U1 on Ascend NPU and optimizes the repeated denoising path using native NPU operators.

## Modifications

- Enable dynamic request batching for SenseNova-U1 on Ascend NPU.
  - Preserve prompt, seed, and output ordering.
  - Require requests in one batch to share resolution, sampling steps, CFG settings, and output options.
  - Add a request-cost estimate based on image tokens, denoising steps, CFG branches, and output count.
  - Keep sequential multi-output requests outside merged dynamic batches.
- Add variable-length prompt support.
  - Pad batched text inputs and track per-sample prefix lengths.
  - Make text and image position indexes batch-aware.
  - Prevent padded prefix tokens from participating in prefix or denoise attention.
  - Reuse the shared unconditional prefix cache for CFG while retaining per-sample conditional prefixes.
  - Generate initial noise from independent per-request RNG streams.
- Add the native Ascend fused inference attention path.
  - Use `torch_npu.npu_fused_infer_attention_score` with per-sample KV lengths.
  - Keep the reusable denoise KV cache in FIA-native BNSD layout.
  - Right-align variable-length prefixes so image tokens keep a common cache offset.
  - Fall back to the existing masked SDPA path when FIA is disabled.
- Optimize repeated NPU denoising work.
  - Use `torch_npu.npu_rms_norm` for RMSNorm.
  - Fuse dense MLP gate/up projections into one matmul followed by `torch_npu.npu_swiglu`.
  - Precompute timestep and noise-scale embeddings outside the denoising loop.
  - Preserve checkpoint-facing MLP parameter names.
- Add offline serving benchmarks, an NPU attention microbenchmark, a bounded-size profiling script, and validation documentation.

The optimized paths are enabled by default on NPU and can be isolated with:

```bash
SGLANG_SENSENOVA_NPU_FIA=0
SGLANG_SENSENOVA_NPU_FUSED_NORM=0
SGLANG_SENSENOVA_NPU_FUSED_MLP=0
```

Dynamic batching remains disabled for SenseNova-U1 on CPU and CUDA. Those platforms retain the existing single-request attention, normalization, MLP, KV-cache, and timestep paths.

Current batching scope:

- Text-to-image generation
- Single Ascend NPU
- One output per merged request
- `think_mode=false` for batches larger than one
- Homogeneous sampling parameters

Image-to-image batching, multi-device execution, CFG conditional/unconditional branch fusion, denoise-step continuous batching, and NPU graph capture are outside this PR.

## Accuracy Tests

The integrated serving path was tested on one Ascend 910C with four concurrent requests.

Smoke-test configuration:

- SenseNova-U1.5-8B-MoT
- 1024 × 1024
- 5 denoising steps
- Batch size 2
- CFG 1 and CFG 4
- Four requests per run

Results:

- All requests completed successfully.
- Server logs confirmed real `2/2` dynamic batches.
- Generated images had the expected dimensions.
- Generated images were visually inspected after enabling fused MLP; no visible quality regression was found.

The native FIA microbenchmark also compared the padded batch output against independently executed compact inputs:

| Comparison | Max absolute error | Mean absolute error | `allclose` |
|---|---:|---:|:---:|
| Masked SDPA, short sample | 0.0009766 | 3.76e-06 | Yes |
| FIA, short sample | 0.0009766 | 1.30e-05 | Yes |
| FIA, long sample | 0.0004883 | 1.29e-05 | Yes |
| Left-aligned FIA cache, short sample | 0.0009766 | 1.30e-05 | Yes |
| Left-aligned FIA cache, long sample | 0.0004883 | 1.29e-05 | Yes |

Tolerance: `atol=0.02`, `rtol=0.02`, BF16.

Unit coverage was added for:

- Per-request seed determinism
- Padded batched tokenization
- Batched position indexes
- Prefix and denoise padding masks
- GQA batch equivalence
- FIA cache alignment
- Fused MLP equivalence and state-dict compatibility
- Scheduler merge/split ordering
- Incompatible request rejection
- NPU-only batching enforcement
- Batched CFG on/off attention behavior

The complete pytest suite was not executed on the Windows development host because that environment does not contain the SGLang PyTorch test dependencies. The test command for CI or an existing Linux inference environment is:

```bash
PYTHONPATH="$PWD/python${PYTHONPATH:+:$PYTHONPATH}" python3 -m pytest \
  python/sglang/multimodal_gen/test/unit/test_sensenova_u1.py -q
```

Static validation completed locally:

- Python compilation
- Ruff error checks
- Ruff formatting
- isort
- codespell
- Git whitespace checks

## Speed Tests and Profiling

Hardware and workload:

- One Ascend 910C
- SenseNova-U1.5-8B-MoT
- Four concurrent requests
- Four generated images
- 2048 × 2048
- 50 denoising steps
- CFG 4
- BF16

| Configuration | Duration (s) | Throughput (images/s) | Mean latency (s) | P95 latency (s) | Peak reserved memory (MB) |
|---|---:|---:|---:|---:|---:|
| B1, optimizations disabled | 234.28 | 0.01707 | 146.47 | 234.27 | 35698 |
| B2, optimizations disabled | 234.83 | 0.01703 | 176.06 | 234.83 | 37870 |
| B1, FIA + RMSNorm + MLP | 205.94 | 0.01942 | 128.80 | 205.94 | 35718 |
| B2, FIA + RMSNorm + MLP | 195.02 | 0.02051 | 146.33 | 195.01 | 37850 |

Derived results:

- B2 without NPU optimizations versus B1 without optimizations: **-0.24% throughput**
- Optimized B1 versus unoptimized B1: **+13.76% throughput**
- Optimized B2 versus unoptimized B2: **+20.42% throughput**
- Optimized B2 versus optimized B1: **+5.60% throughput**
- Optimized B2 versus unoptimized B1: **+20.13% throughput**
- Optimized B2 reduced total duration by **16.76%** versus unoptimized B1.

The attention microbenchmark at query length 4096 measured:

| Attention path | Median latency |
|---|---:|
| B2 unmasked SDPA | 5.060 ms |
| B2 masked SDPA | 6.544 ms |
| B2 native FIA | 4.718 ms |

Native FIA was:

- **1.387× faster** than masked SDPA.
- **1.073× faster** than unmasked SDPA.
- **1.103× higher theoretical B2 throughput** than the measured B1 SDPA path.

Profiling showed that the main end-to-end gain comes from the NPU attention, RMSNorm, and MLP operator paths. Request batching adds another 5.60% throughput on top of those optimizations.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #34586963431](https://github.com/sgl-project/sglang/actions/runs/34586963431)
Latest PR Test (Extra): :x: [Run #34586963191](https://github.com/sgl-project/sglang/actions/runs/34586963191)
Latest PR Test (AMD ROCm 10): :x: [Run #34586963246](https://github.com/sgl-project/sglang/actions/runs/34586963246)

### CODE REVIEW

File: benchmark/sensenova/bench_npu_attention.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+import argparse
+import json
+import math
+import statistics
+import time
+from pathlib import Path
+
+
+def main():
```
Comment (ping1jing2) [inline benchmark/sensenova/bench_npu_attention.py:11]: This code is a bit hard to maintain as-is, could you refactor int into a few smaller func? thinks.

File: benchmark/sensenova/bench_npu_attention.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+                right_padded_k,
+                right_padded_v,
+                actual_seq_lengths=actual_seq_lengths,
+                actual_seq_lengths_kv=actual_seq_lengths_kv,
+                num_heads=args.heads,
+                num_key_value_heads=args.kv_heads,
+                scale=scale,
+                input_layout="BNSD",
+                sparse_mode=0,
+            )
```
Comment (ping1jing2) [inline benchmark/sensenova/bench_npu_attention.py:143]: one more questions here, could you please tell me the purpose of this file. we already had responding test file in https://github.com/Ascend/op-plugin/blob/master/test/test_custom_ops/test_npu_fused_infer_attention_score.py

File: docs/docs/developer_guide/sensenova-u1-npu-batching.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
+---
```
Comment (ping1jing2) [inline docs/docs/developer_guide/sensenova-u1-npu-batching.md:1]: could you also update the Ascend result in cookbook/diffusion/SenseNova

File: python/sglang/multimodal_gen/configs/pipeline_configs/sensenova_u1.py [hunk 4]
```diff
@@ -8,6 +8,10 @@
 from sglang.multimodal_gen.configs.pipeline_configs.model_deployment_config import (
     ModelDeploymentConfig,
 )
+from sglang.multimodal_gen.configs.sensenova_u1 import (
+    SENSENOVA_U1_RESOLUTION_ALIGNMENT,
+)
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/configs/pipeline_configs/sensenova_u1.py:13]: shall we rename it to `RESOLUTION_ALIGNMENT`. because we define it in the `sensenova_u1.py`, then it's noneeed to add model name as prefix.

File: python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py [hunk 5]
```diff
@@ -55,6 +57,30 @@
...
+        "off",
+    )
+
+
+def npu_fused_mlp_enabled() -> bool:
+    return os.getenv("SGLANG_SENSENOVA_NPU_FUSED_MLP", "1").lower() not in (
+        "0",
+        "false",
+        "off",
+    )
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py:81]: could you add it into `python\sglang\multimodal_gen\envs.py`

File: python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py [hunk 6]
```diff
@@ -146,44 +180,121 @@ def _sdpa_attn_func(
...
+            num_key_value_heads = k.shape[1]
+        elif input_layout == "BSND":
+            q_bhsd = q.transpose(1, 2).contiguous()
+            k_bhsd = k.transpose(1, 2).contiguous()
+            v_bhsd = v.transpose(1, 2).contiguous()
+            batch_size, query_length, num_heads, _ = q.shape
+            num_key_value_heads = k.shape[2]
+        else:
+            raise ValueError(f"Unsupported attention input layout: {input_layout}")
+        out, _ = torch_npu.npu_fused_infer_attention_score(
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py:232]: please call it in the way of `torch.ops.npu.npu_fused_infer_attention_score`, then we don't need to `import torch_npu`. please check other kernels too

**Review by ping1jing2 — COMMENTED** [review]
this PR is too huge to review it. i didn't review all the details...

**Comment by ping1jing2** [comment]
@Makcum888e @OrangeRedeng please help to review this PR when you are free

File: benchmark/sensenova/NPU_ATTENTION_MICROBENCH.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
+# SenseNova NPU Attention 微基准
+
+这个测试绕过服务和完整模型，直接调用 SenseNova 当前 NPU 路径使用的
```
Comment (ChefWu551) [inline benchmark/sensenova/NPU_ATTENTION_MICROBENCH.md:3]: Please  describe in English or it will make lint fail

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/sensenova_u1/stages/generation.py [hunk 8]
```diff
@@ -71,7 +72,47 @@ def forward(self, batch: Req, server_args: ServerArgs) -> OutputBatch:
                 "SenseNova-U1 expects output expansion before generation; "
                 f"got num_outputs_per_prompt={batch.num_outputs_per_prompt}."
             )
-        seed = batch.seed[0] if isinstance(batch.seed, list) else int(batch.seed)
+        if current_platform.is_npu():
```
Comment (ChefWu551) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/sensenova_u1/stages/generation.py:75]: This batch process seems can be work in GPU too. Could you explain why only NPU is supported at the moment?

File: benchmark/sensenova/NPU_PROFILING.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
+# SenseNova B1/B2 短步骤 NPU profiling
+
```
Comment (ChefWu551) [inline benchmark/sensenova/NPU_PROFILING.md:2]: It looks like a lot of intermediate files have been included in this PR. Could we remove the unnecessary intermediate files and keep only the final result/documentation files.

File: docs/docs/developer_guide/sensenova-u1-npu-batching.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+title: "SenseNova-U1 NPU batching development plan"
```
Comment (ChefWu551) [inline docs/docs/developer_guide/sensenova-u1-npu-batching.md:2]: Consider adding a  Advanced Usage (like QwenImage) section to the documentation to explain how to use this feature. Could we remove useless part. This Doc seems not belong here

File: python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py
Comment (Makcum888e) [inline python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py:335]: could we use `from sglang.srt.layers.layernorm import RMSNorm` instead of reimplementing ir again?

File: python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py [hunk 11]
```diff
@@ -227,8 +345,52 @@ def __init__(self, config):
...
+        return hasattr(torch_npu, "npu_swiglu")
 
     def forward(self, x):
+        if self._use_npu_fused_mlp(x):
+            import torch_npu
+
+            gate_up = F.linear(x, self._pack_npu_gate_up_weights())
+            return self.down_proj(torch_npu.npu_swiglu(gate_up))
         down_proj = self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
         return down_proj
```
Comment (Makcum888e) [inline python/sglang/multimodal_gen/runtime/models/sensenova_u1/neo_unify/modeling_qwen3.py:395]: please, check if we can use `Qwen3MLP` or more classes from `sglang.multimodal_gen.runtime.models.encoders.qwen3`
